### PR TITLE
[WIP] Memory manager reusables

### PR DESCRIPTION
**Pull request by k-wiatrzyk**
TODO list:
- [ ] update existing tests
- [x] add more test cases 
- [x] logic for flushing `memoryToReuse`
- [x] error handling for `updateMemoryToReuse()`
- [ ] describe problematic edge cases
- [x] remove debug logging

Challenges:
1. InitContainer reserved resources on NUMA 0 [01-true]. AppContainers requires higher amount of resources than InitContainer and there is no enough on NUMA 0 (including inheriting from InitContainer). NUMA 1 has enough of resources to fill the requirement but it must inherit from InitContainer. Thus instead of producing hints ([10-true]) it will produce none (empty resource hint). Because in Memory Manager when resources were allocated to one of nodes, multi numa group can't be produced ([11-false]) thus, pod will be rejected. This problem only exist in container scope.
2. When container crashes and is recreated should he also inherit resources from init containers? 
3. TBD

### CODE REVIEW

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 1]
```diff
@@ -45,6 +45,8 @@ type staticPolicy struct {
 	systemReserved systemReservedMemory
 	// topology manager reference to get container Topology affinity
 	affinity topologymanager.Store
+	// init container's memory and hugepages reservation that can be reused by app containers in the pod
+	memoryToReuse map[string]map[v1.ResourceName]state.Block
```
Comment (k-wiatrzyk) [inline pkg/kubelet/cm/memorymanager/policy_static.go:49]: It could be also a `map[string][]state.Block` to not duplicate `Block.Type` value but making it a nested map makes logic in few places simpler (without nested loops).

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 2]
```diff
@@ -83,6 +86,46 @@ func (p *staticPolicy) Start(s state.State) error {
 	return nil
 }
 
+func (p *staticPolicy) updateMemoryToReuse(pod *v1.Pod, container *v1.Container, blocks []state.Block){
+	// This will flush reusable resources for a pod after all containers have been allocated
+	// for podUID := range p.memoryToReuse {
+	// 	if podUID != string(pod.UID) {
+	// 		delete(p.memoryToReuse, podUID)
+	// 	}
+	// }
```
Comment (k-wiatrzyk) [inline pkg/kubelet/cm/memorymanager/policy_static.go:95]: Commented right now.

The same logic is applied in CPU Manager. It means that reusable for a pod will be deleted when all containers have been allocated resources and new pod is being allocated right now.

The problem I see here: what happens when after a couple of hours and few other pods placed a container crashes. (there may be a few seconds, minutes of BackOff), shouldn't it use the reusables? Can there be a racing for that memory with other containers? 

Here important may be a sequence (RemoveContainer(), AddContainer(), etc) when container crashes to understand that.

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 3]
```diff
@@ -83,6 +86,46 @@ func (p *staticPolicy) Start(s state.State) error {
...
+
+	//Otherwise it is an app container.
+	//Remove its memory and hugepages from the reusable pool for any new allocations.
+	for _, block := range blocks {
+		if _, ok := p.memoryToReuse[string(pod.UID)]; !ok {
+			continue
+		} 
+		reusableBlock := p.memoryToReuse[string(pod.UID)][block.Type]
+		reusableBlock.Size -= block.Reused
+		p.memoryToReuse[string(pod.UID)][block.Type] = reusableBlock
```
Comment (k-wiatrzyk) [inline pkg/kubelet/cm/memorymanager/policy_static.go:125]: It lacks checking edge cases (i.e. what if `block.Reused` > `reusableBlock.Size`?), right now it will produce not nice very high value (edge value of `uint64). I added to checklist to add error handling for this func).

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 4]
```diff
@@ -109,7 +154,7 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
 	// topology manager returned the hint with NUMA affinity nil
 	// we should use the default NUMA affinity calculated the same way as for the topology manager
 	if hint.NUMANodeAffinity == nil {
-		defaultHint, err := p.getDefaultHint(s, requestedResources)
+		defaultHint, err := p.getDefaultHint(s, requestedResources, string(pod.UID))
```
Comment (k-wiatrzyk) [inline pkg/kubelet/cm/memorymanager/policy_static.go:157]: Strange change, I haven't done that. It must have lasted refactoring. Will drop it.

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 5]
```diff
@@ -233,6 +310,11 @@ func (p *staticPolicy) RemoveContainer(s state.State, podUID string, containerNa
 		}
 	}
 
+	// If this was a last allocation from pod free reusables
+	if len(s.GetMemoryAssignments()[podUID]) == 0 {
+		delete(p.memoryToReuse, podUID)
+	}
```
Comment (k-wiatrzyk) [inline pkg/kubelet/cm/memorymanager/policy_static.go:316]: `Size` in reusables is equal to all `Size`s in initContainer blocks, so we don't have to explicitly update NodeMap after deleting `memoryToReuse` entry, it was correctly updated when initContainers were deleted

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 6]
```diff
@@ -707,7 +818,7 @@ func isAffinitySatisfyRequest(machineState state.NodeMap, mask bitmask.BitMask,
 // it possible that we will get the subset of hint that we provided to the topology manager, in this case we want to extend
 // it to the original one
 func (p *staticPolicy) extendTopologyManagerHint(s state.State, requestedResources map[v1.ResourceName]uint64, mask bitmask.BitMask) (*topologymanager.TopologyHint, error) {
-	hints := p.calculateHints(s, requestedResources)
+	hints := p.calculateHints(s, requestedResources, nil)
```
Comment (k-wiatrzyk) [inline pkg/kubelet/cm/memorymanager/policy_static.go:821]: This should be updated.
#TODO

File: pkg/kubelet/cm/memorymanager/state/state.go [hunk 7]
```diff
@@ -83,6 +83,7 @@ type Block struct {
 	NUMAAffinity []int           `json:"numaAffinity"`
 	Type         v1.ResourceName `json:"type"`
 	Size         uint64          `json:"size"`
+	Reused		 uint64	         `json:"reused`
```
Comment (k-wiatrzyk) [inline pkg/kubelet/cm/memorymanager/state/state.go:86]: `Reused` stores amount of memory that was reused by app or init container from the `memoryToReuse` pool.

**Comment by cynepco3hahue** [comment]
> 2\. When container crashes and is recreated should he also inherit resources from init containers?

If we are talking only about the container restart and not about the whole pod, it should just regenerate hints.

**Comment by k-wiatrzyk** [comment]
Ah, right. Because container name is used there in mapping, not container ID (which is different when container restarts). This piece was missing in my head.

**Comment by cynepco3hahue** [comment]
> 1. InitContainer reserved resources on NUMA 0 [01-true]. AppContainers requires higher amount of resources than InitContainer and there is no enough on NUMA 0 (including inheriting from InitContainer). NUMA 1 has enough of resources to fill the requirement but it must inherit from InitContainer. Thus instead of producing hints ([10-true]) it will produce none (empty resource hint). Because in Memory Manager when resources were allocated to one of nodes, multi numa group can't be produced ([11-false]) thus, pod will be rejected. This problem only exist in container scope.

I think we should decide what use cases for the init container memory re-use is an important one. I thought it mostly needed to increase the density of pods on the NUMA nodes. Because the allocation of memory resources happens before a container start, so in general you do not want to reject the pod when `init container memory + app containers memory > NUMA node free memory`, but `app containers memory <= NUMA node free memory`. If it the only use case, I think we should not reject the hint [10-true]. But if it some use case when it really important to be sure that app container will run exactly on the same NUMA node where init container ran, we will need definitely to reject the [10-true] hint.

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 8]
```diff
@@ -121,11 +166,31 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
...
+
+	for resourceName, resourceSize := range requestedResources {
+		//copy a map
+		requestedAbsolute[resourceName] = resourceSize
+
+		if _, ok := reusable[resourceName]; !ok {
+			continue
+		} 
+
+		reusableSize := reusable[resourceName].Size
```
Comment (cynepco3hahue) [inline pkg/kubelet/cm/memorymanager/policy_static.go:181]: like you wrote above we should be sure that `reusable[resourceName].Size` > 0

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 9]
```diff
@@ -121,11 +166,31 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
...
+
+		if resourceSize > reusableSize {
+			requestedAbsolute[resourceName] = resourceSize - reusableSize
+			continue
+		}
+		requestedAbsolute[resourceName] = 0
```
Comment (cynepco3hahue) [inline pkg/kubelet/cm/memorymanager/policy_static.go:187]: should we reduce `reusableSize` by the resourceSize?
something like that
```golang
p.memoryToReuse[string(pod.UID)][resourceName].Size -= resourceSize
```
or updating it under https://github.com/mgjeong/kubernetes/pull/30/files#diff-4c005e5ddcd048f3c5db48df4fdf1fa4ae7318d0d54f3c0964d074910946a2d1R251 should be enough

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 10]
```diff
@@ -196,6 +266,13 @@ func (p *staticPolicy) RemoveContainer(s state.State, podUID string, containerNa
 	// Mutate machine memory state to update free and reserved memory
 	machineState := s.GetMachineState()
 	for _, b := range blocks {
+		// This container reused some resources and they must be released to the reusable pool
+		if b.Reused != 0 {
```
Comment (cynepco3hahue) [inline pkg/kubelet/cm/memorymanager/policy_static.go:270]: I unsure we need it, if the container crashed, the allocation should be regenerated from the state file

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 11]
```diff
@@ -382,7 +469,7 @@ func getRequestedResources(container *v1.Container) (map[v1.ResourceName]uint64,
 	return requestedResources, nil
 }
 
-func (p *staticPolicy) calculateHints(s state.State, requestedResources map[v1.ResourceName]uint64) map[string][]topologymanager.TopologyHint {
+func (p *staticPolicy) calculateHints(s state.State, requestedResources map[v1.ResourceName]uint64, reusable map[v1.ResourceName]state.Block) map[string][]topologymanager.TopologyHint {
```
Comment (cynepco3hahue) [inline pkg/kubelet/cm/memorymanager/policy_static.go:472]: we should check that `reusable != nil` before use it

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 12]
```diff
@@ -403,6 +492,28 @@ func (p *staticPolicy) calculateHints(s state.State, requestedResources map[v1.R
 			return
 		}
 
+		for resourceName, resourceSize := range requestedResources {
+			//copy a map
```
Comment (cynepco3hahue) [inline pkg/kubelet/cm/memorymanager/policy_static.go:496]: nit: space

**Comment by k-wiatrzyk** [comment]
I'm not sure if I describe it correctly. Hint [10-true] will be rejected not because of `init container memory + app containers memory > NUMA node free memory` but because InitContainer is on [01-true], because it was allocated first and AppContainer **must** inherit that. And if we would place an AppContainer on [10-true] we have the following issue:
1. Scheduler sees that memory occupied by this pod is equal to **Effective limits/requests** = _max(max(InitContainers),sum(AppContainers))_
2. Memory on the node will be freed (in **_NodeMap_**) after container failure or whole pod removal (I mean calling `RemoveContainer()` here) 
3. So we won't free memory used by **InitContainers** and because of that, this pod, will occupy more memory in **_NodeMap_** than **Effective limits/requests** assumes. Thus, issues with scheduling can occur.

**Comment by cynepco3hahue** [comment]
> **Solution for that?**
> If we want to allocate resource for AppContainer on [10-true] then, when `AddContainer` is called (so container is starting), we check if this container is app container or not. If it is an app container, then free all resources that were allocated for Init Containers and are not reused by App Containers. What do you think about such solution?

IMHO we can just uncomment `This will flush reusable resources for a pod after all containers have been allocated` and it will work, WDYT?

**Comment by k-wiatrzyk** [comment]
> IMHO we can just uncomment `This will flush reusable resources for a pod after all containers have been allocated` and it will work, WDYT?

To be honest, I'm not sure. I agree that it should be uncommented, as I haven't properly understood the logic of regenerating hints. But uncommenting this won't solve the problem I described because it will flush only `memoryToReuse` and not `NodeMap`. 

Because after InitContainer executed successfully it's allocation will exist in **NodeMap** until `RemoveContainer()` will be called. And that allocation could be cleared only when InitContainer executed successfully, not earlier. 

From Slack, regarding CPU Manager, same problem:
> It will involve allowing init containers to release CPUs back to the defaultCPUSet if-and-only-if app containers don’t inherit from them (and thus also allow app containers to not inherit from them under certain conditions).
> Right now, no CPUs are released back to the defaultCPUSet until the entire pod is shutdown

File: pkg/kubelet/cm/memorymanager/policy_static.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go:49]: > When we are talking about the topology manager container policy, it possible that memory allocation for containers inside of the same pod will be done from different NUMA nodes, and I do not see any checks related to where the block was allocated. Please point me to the relevant checks in case if I missed it.
> 
> For example, I can have two init containers, the memory manager allocated 250Mi of the memory for the first init container from the NUMA node 0, and for the second init container allocated 500Mi from the NUMA node 1. So the amount of re-usable memory for the pod is different for each NUMA node.

IIUC, right now, InitContainer2 must Inherit resources, thus the same NUMA node, as InitContainer1. So there won't be a situation, when you have two InitContainers on different NUMA nodes, regardless Topology Manager policy.

[the remark this answers is not in the document]
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go:49]: According to your example and implementation in this PR:
- second init container will inherit 250Mi from first init container and will allocate additional 250Mi from the same NUMA node
- it can't be placed entirely on different NUMA node

File: pkg/kubelet/cm/memorymanager/policy_static.go (hunk 2 above)
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go]: That's right. Thanks for writing it down. I uncommented that part and deleted logic from `RemoveContainer()`

File: pkg/kubelet/cm/memorymanager/policy_static.go (hunk 6 above)
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go]: Done

File: pkg/kubelet/cm/memorymanager/state/state.go (hunk 7 above)
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/state/state.go:86]: True, but adding it here make sense for me as it is a container that reuses the memory. I also thought about accountability, It may be easy to check if a container reused memory by checking state file, right?

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 13]
```diff
@@ -83,6 +86,46 @@ func (p *staticPolicy) Start(s state.State) error {
...
+				}
+				reusableBlock := p.memoryToReuse[string(pod.UID)][block.Type]
+				reusableBlock.Size += block.Size
+				p.memoryToReuse[string(pod.UID)][block.Type] = reusableBlock
+			}
+			return
+		}
+	}
+
+	//Otherwise it is an app container.
```
[the remark this answers is not in the document]
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go]: Done.

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 14]
```diff
@@ -83,6 +86,46 @@ func (p *staticPolicy) Start(s state.State) error {
...
+	//Remove its memory and hugepages from the reusable pool for any new allocations.
+	for _, block := range blocks {
+		if _, ok := p.memoryToReuse[string(pod.UID)]; !ok {
+			continue
```
[the remark this answers is not in the document]
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go:124]: True, thanks. Done.

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 15]
```diff
@@ -83,6 +86,46 @@ func (p *staticPolicy) Start(s state.State) error {
...
+		} 
+		reusableBlock := p.memoryToReuse[string(pod.UID)][block.Type]
```
[the remark this answers is not in the document]
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go:126]: Yes, done.

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 16]
```diff
@@ -121,11 +166,31 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
...
 	machineState := s.GetMachineState()
+	reusable := p.memoryToReuse[string(pod.UID)]
+	// requested memory - reusable 
+	requestedAbsolute := make(map[v1.ResourceName]uint64)
+
+	for resourceName, resourceSize := range requestedResources {
+		//copy a map
+		requestedAbsolute[resourceName] = resourceSize
+
+		if _, ok := reusable[resourceName]; !ok {
```
[the remark this answers is not in the document]
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go]: Right, good catch. Done.

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 17]
```diff
@@ -121,11 +166,31 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
 	}
 
 	machineState := s.GetMachineState()
+	reusable := p.memoryToReuse[string(pod.UID)]
+	// requested memory - reusable 
+	requestedAbsolute := make(map[v1.ResourceName]uint64)
+
+	for resourceName, resourceSize := range requestedResources {
+		//copy a map
```
[the remark this answers is not in the document]
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go]: Done.

File: pkg/kubelet/cm/memorymanager/policy_static.go (hunk 8 above)
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go:191]: This is checked now in `updateMemoryToReuse()`. Should it be checked here too?

File: pkg/kubelet/cm/memorymanager/policy_static.go (hunk 9 above)
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go:197]: `p.updateMemoryToReuse()` do the job

File: pkg/kubelet/cm/memorymanager/policy_static.go (hunk 10 above)
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go]: Removed logic from `RemoveContainer()`

File: pkg/kubelet/cm/memorymanager/policy_static.go (hunk 12 above)
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go:503]: done.

File: pkg/kubelet/cm/memorymanager/policy_static.go (hunk 11 above)
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go:479]: true, there are checks now in proper places.

**Review by cynepco3hahue — COMMENTED** [review]
Some general comments
- run `make quick-verify` to be sure that all linters passed
- update the state test to have a valid checksum(https://github.com/mgjeong/kubernetes/blob/190436b1155c3f181e391c4df0d13fa2603ff9fa/pkg/kubelet/cm/memorymanager/state/state_checkpoint_test.go#L68) to `653911231`

File: pkg/kubelet/cm/memorymanager/policy_static_test.go [hunk 18]
```diff
@@ -2340,3 +2344,1724 @@ func TestStaticPolicyGetTopologyHints(t *testing.T) {
...
+					hugepages1Gi: state.Block{
+						NUMAAffinity: []int{0},
+						Type:         hugepages1Gi,
+						Size:         4 * gb,
+					},
+				},
+			},
+		},
+		{
+			description: "app containers request higher than reusable, should allocate from reusable and from nodemap",
```
Comment (cynepco3hahue) [inline pkg/kubelet/cm/memorymanager/policy_static_test.go:2925]: Does it the same test as https://github.com/mgjeong/kubernetes/pull/30/files#diff-7ddae4e94b2576ee01e6b475370eb488a637d80bbe6f391ae6cdfd128311811aR3109?

File: pkg/kubelet/cm/memorymanager/policy_static_test.go [hunk 19]
```diff
@@ -2340,3 +2344,1724 @@ func TestStaticPolicyGetTopologyHints(t *testing.T) {
...
+					},
+					{
+						NUMANodeAffinity: newNUMAAffinity(0, 1),
+						Preferred:        false,
+					},
+				},
+			},
+			memoryToReuse: map[string]map[v1.ResourceName]state.Block{},
+			expectedMemoryToReuse: map[string]map[v1.ResourceName]state.Block{
+				"pod2": map[v1.ResourceName]state.Block{
```
Comment (cynepco3hahue) [inline pkg/kubelet/cm/memorymanager/policy_static_test.go:2519]: nit: you can remove `map[v1.ResourceName]state.Block` the same true for all tests

File: pkg/kubelet/cm/memorymanager/policy_static_test.go [hunk 20]
```diff
@@ -2340,3 +2344,1724 @@ func TestStaticPolicyGetTopologyHints(t *testing.T) {
...
+					"memory": state.Block{
```
Comment (cynepco3hahue) [inline pkg/kubelet/cm/memorymanager/policy_static_test.go:2520]: nit: you can remove `state.Block`, the same true for all tests

File: pkg/kubelet/cm/memorymanager/policy_static_test.go [hunk 21]
```diff
@@ -2340,3 +2344,1724 @@ func TestStaticPolicyGetTopologyHints(t *testing.T) {
...
+						Size:         1 * gb,
+					},
+				},
+			},
+		},
+		{
+			description: "generate hints and allocate second init container when one was already allocated",
+			assignments: state.ContainerMemoryAssignments{
+				"pod2": map[string][]state.Block{
+					"container1": {
```
Comment (cynepco3hahue) [inline pkg/kubelet/cm/memorymanager/policy_static_test.go:2537]: nit: initcontainer

File: pkg/kubelet/cm/memorymanager/policy_static_test.go [hunk 22]
```diff
@@ -2340,3 +2344,1724 @@ func TestStaticPolicyGetTopologyHints(t *testing.T) {
...
+							NUMAAffinity: []int{0},
+							Type:         hugepages1Gi,
+							Size:         1 * gb,
+						},
+					},
+				},
+			},
+			expectedAssignments: state.ContainerMemoryAssignments{
+				"pod2": map[string][]state.Block{
+					"container1": {
```
Comment (cynepco3hahue) [inline pkg/kubelet/cm/memorymanager/policy_static_test.go:2553]: nit: initcontainer

File: pkg/kubelet/cm/memorymanager/policy_static_test.go [hunk 23]
```diff
@@ -2340,3 +2344,1724 @@ func TestStaticPolicyGetTopologyHints(t *testing.T) {
...
+							Type:         v1.ResourceMemory,
+							Size:         1 * gb,
+						},
+						{
+							NUMAAffinity: []int{0},
+							Type:         hugepages1Gi,
+							Size:         1 * gb,
+						},
+					},
+					"container2": {
```
Comment (cynepco3hahue) [inline pkg/kubelet/cm/memorymanager/policy_static_test.go:2565]: nit: initcontainer2

File: pkg/kubelet/cm/memorymanager/policy_static_test.go [hunk 24]
```diff
@@ -2340,3 +2344,1724 @@ func TestStaticPolicyGetTopologyHints(t *testing.T) {
...
+					NumberOfAssignments: 0,
+				},
+			},
+			pod: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					UID: types.UID("pod2"),
+				},
+				Spec: v1.PodSpec{
+					InitContainers: []v1.Container{
+						{
```
Comment (cynepco3hahue) [inline pkg/kubelet/cm/memorymanager/policy_static_test.go:2671]: nit: initcontainer2

File: pkg/kubelet/cm/memorymanager/policy_static_test.go [hunk 25]
```diff
@@ -2340,3 +2344,1724 @@ func TestStaticPolicyGetTopologyHints(t *testing.T) {
...
+						Size:         1 * gb,
+					},
+				},
+			},
+			expectedMemoryToReuse: map[string]map[v1.ResourceName]state.Block{
+				"pod2": map[v1.ResourceName]state.Block{
+					"memory": state.Block{
+						NUMAAffinity: []int{0},
+						Type:         v1.ResourceMemory,
+						Size:         2 * gb,
```
Comment (cynepco3hahue) [inline pkg/kubelet/cm/memorymanager/policy_static_test.go:2730]: I think we should not sum, we should use the max value from all init containers

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 26]
```diff
@@ -83,6 +86,54 @@ func (p *staticPolicy) Start(s state.State) error {
...
+			if _, ok := p.memoryToReuse[string(pod.UID)]; !ok {
+				p.memoryToReuse[string(pod.UID)] = make(map[v1.ResourceName]state.Block)
+			}
+			for _, block := range blocks {
+				if _, ok := p.memoryToReuse[string(pod.UID)][block.Type]; !ok {
+					p.memoryToReuse[string(pod.UID)][block.Type] = block
+					continue
+				}
+				reusableBlock := p.memoryToReuse[string(pod.UID)][block.Type]
+				reusableBlock.Size += block.Size
```
[the remark this answers is not in the document]
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go:110]: Yes, you are right, working on that :)

**Comment by k-wiatrzyk** [comment]
FYI: still checking some stuff and need to address few comments
TODO: Add multi-numa test cases

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 27]
```diff
@@ -83,6 +86,54 @@ func (p *staticPolicy) Start(s state.State) error {
...
+	// Remove its memory and hugepages from the reusable pool for any new allocations.
+	for _, block := range blocks {
+		if _, ok := p.memoryToReuse[string(pod.UID)]; !ok {
+			break
+		}
+		if _, ok := p.memoryToReuse[string(pod.UID)][block.Type]; !ok {
+			continue
+		}
+		reusableBlock := p.memoryToReuse[string(pod.UID)][block.Type]
+		if block.Reused > reusableBlock.Size {
```
[the remark this answers is not in the document]
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go:127]: No, we should not. Returning an error from this place by this func will cause `Allocate()` to throw an error, thus rejecting the pod.

File: pkg/kubelet/cm/memorymanager/state/state.go (hunk 7 above)
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/state/state.go]: Thanks, I corrected it. 
This is how it looks right now: 
```
	// Reused contains amount of memory common with init container's allocation
	// Reused + Size is equal to amount of container's requirement of that resource
	Reused uint64 `json:"reused"`
```

File: pkg/kubelet/cm/memorymanager/policy_static.go (hunk 12 above)
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go]: Oh, my fault. Now it is okay :)

File: pkg/kubelet/cm/memorymanager/policy_static.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go:110]: You can check this commit for that functionality: https://github.com/mgjeong/kubernetes/pull/30/commits/117af55a66d9552dc3fa2d6a93dc9810accfc2cd

File: pkg/kubelet/cm/memorymanager/policy_static.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go:127]: And it shouldn't happen actually.

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 28]
```diff
@@ -92,6 +143,10 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
 
 	klog.Infof("[memorymanager] Allocate (pod: %s, container: %s)", pod.Name, container.Name)
 	if blocks := s.GetMemoryBlocks(string(pod.UID), container.Name); blocks != nil {
+		// TODO: reusable check
```
[the remark this answers is not in the document]
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go]: great, removed

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 29]
```diff
@@ -180,6 +257,11 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
 	s.SetMachineState(machineState)
 	s.SetMemoryBlocks(string(pod.UID), container.Name, containerBlocks)
 
+	//update reusable resources
```
[the remark this answers is not in the document]
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go]: Done.

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 30]
```diff
@@ -121,11 +176,31 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
...
 	machineState := s.GetMachineState()
+	reusable := p.memoryToReuse[string(pod.UID)]
+	// requested memory - reusable
+	requestedAbsolute := make(map[v1.ResourceName]uint64)
+
+	for resourceName, resourceSize := range requestedResources {
+		// copy a map
+		requestedAbsolute[resourceName] = resourceSize
+
+		if _, ok := reusable[resourceName]; !ok || reusable == nil {
```
[the remark this answers is not in the document]
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go]: I actually checked it and the second condition is not needed. 
Please, take a look at: https://play.golang.org/p/iwWIsPV2hW1 
Is it a correct example?

File: pkg/kubelet/cm/memorymanager/policy_static_test.go (hunk 18 above)
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static_test.go]: They were different:
- in first init container has hugepages that can be reused
- in second it doesn't 
I updated descriptions to be more precise. If you think that one of those tests is redundant I can delete it. (I wonder if it is)

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 31]
```diff
@@ -350,7 +435,20 @@ func (p *staticPolicy) GetTopologyHints(s state.State, pod *v1.Pod, container *v
...
 	}
 
-	requestedResources, err := getRequestedResources(container)
+	var requestedResources map[v1.ResourceName]uint64
+	var err error
+
+	// Init containers should be placed on NUMA nodes where inheriting container's requirement can be satisfied
+	for _, initContainer := range pod.Spec.InitContainers {
+		if container.Name == initContainer.Name {
+			requestedResources, err = getMaxRequestedResources(pod)
```
Comment (cynepco3hahue) [inline pkg/kubelet/cm/memorymanager/policy_static.go:444]: I think you should not run it for each init container, it should be enough to run it only once and save the value under some object variable

**Review by cynepco3hahue — COMMENTED** [review]
I think we should:
- calculate once for each pod init containers maximum request for all memory resources and save it under `memoryToReuse`
- once we calculate hints, verify if **the free memory on the NUMA node + `memoryToReuse`(when init containers started on this NUMA node) >= app requested memory**, if yes, use this hint, if not pass to the next NUMA node
In the worst case, we will not re-use the memory at all, but it not so bad because the memory allocated by init containers will be released once the init container is removed.

I do not really want to calculate maximum request for all containers, because it does not really solve the problem and creates a new one:
- old problem: we have memory to reuse, but the NUMA node does not have enough memory to start the app container. Because of the calculation of the max requester for all containers, you can solve the problem for the first app container, but no one guarantee that it will work for the second app container. For example, the NUMA node has 300Mi of free memory and your max request was 250Mi, so it succeeded to start the first container that requested 200Mi(you still have 50Mi to re-use) but will fail to start the second one that requested 150Mi(200Mi + 150Mi=350Mi)
- new problem: it possible that we have two app containers and one requested a lot of memory and the second one a lot of hugepages, but under the system, you have only one NUMA node with a lot of hugepages, but it does not have enough memory. So the container will be rejected because we calculate the maximum for both containers.

**Comment by k-wiatrzyk** [comment]
1
-
> * calculate once for each pod init containers maximum request for all memory resources and save it under `memoryToReuse`
> * once we calculate hints, verify if **the free memory on the NUMA node + `memoryToReuse`(when init containers started on this NUMA node) >= app requested memory**, if yes, use this hint, if not pass to the next NUMA node
>   In the worst case, we will not re-use the memory at all, but it not so bad because the memory allocated by init containers will be released once the init container is removed.

I considered this. But have in mind that init container allocation will be deleted on `RemoveContainer()`, which IIRC is called only when the whole pod is removed. Thus, the problem with scheduling. 

For such example, you described above, the following situation will happen:
1. Pod to schedule (let's call it TestPod): 
- init container: 2 Gi
- app container: 3 Gi
2. Scheduler calculates Effective resources requests/limits: 3 Gi
3. Scheduler selects Node01 (free 6 Gi - numa0: 2Gi, numa1: 4Gi). Node01 has 3 Gi free visible by Scheduler after scheduling TestPod to it.
4. TestPod is allocated:
- init container: numa0 (left 0 Gi free)
- app container: numa1 (left 1 Gi free)
5. So Scheduler sees that Node01 has 3Gi free but in fact, it has only 1Gi left. 

I am wondering if it is acceptable. I would really like to have a solution that will address all cases :/ 

2
-

> I do not really want to calculate maximum request for all containers, because it does not really solve the problem and creates a new one:
> 
> * old problem: we have memory to reuse, but the NUMA node does not have enough memory to start the app container. Because of the calculation of the max requester for all containers, you can solve the problem for the first app container, but no one guarantee that it will work for the second app container. For example, the NUMA node has 300Mi of free memory and your max request was 250Mi, so it succeeded to start the first container that requested 200Mi(you still have 50Mi to re-use) but will fail to start the second one that requested 150Mi(200Mi + 150Mi=350Mi)
> * new problem: it possible that we have two app containers and one requested a lot of memory and the second one a lot of hugepages, but under the system, you have only one NUMA node with a lot of hugepages, but it does not have enough memory. So the container will be rejected because we calculate the maximum for both containers.

Alright, I agree that above issues are also existing right now. I also thought about first one at Friday. But I didn't realized that second exists. So true, taking max of all containers is not a good idea.
